monai/data/wsi_datasets.py

```diff
@@ -256,7 +256,7 @@ def __init__(
         self.mask_level = mask_level
         # Create single sample for each patch (in a sliding window manner)
         self.data = []
```

Comment:
Hi @drbeh ,

Thanks for the quick fix, seems this line `self.data = []` is not necessary anymore?
And I still don't quite understand what exact issue the previous code has? Could you please help give a simple example or code link?

Thanks in advance.

Reply:
Hi @Nic-Ma,

Thank for reviewing this. You are right `self.data = []` is not necessary for the code functionality but I left it there for type checking (since I am using .extend and it doesn't work with any Sequence). I could just add `self.data: List` but I thought it is more readable to explicitly make it an empty list.

To explain the issue, let me explain what is happening to `self.image_data` first. A `sample` which is one entry of `self.image_data` is being passed to `_evaluate_path_location` to not only create `self.data` but also to get enriched by new information: https://github.com/Project-MONAI/MONAI/blob/f7f3461b84e55963ef49fd7926825fc1d5c1cad5/monai/data/wsi_datasets.py#L394-L398
The second part that is adding new information to `sample` depends on `self.image_data` being a mutable so that change `sample` updates `self.image_data`. This issue was that if the input `data` is immutable, `self.image_data` becomes immutable and updating `sample` not longer updates `self.image_data`.

Reply:
> Hi @Nic-Ma,
> 
> Thanks for reviewing this. You are right `self.data = []` is not necessary for the code functionality but I left it there for type checking (since I am using .extend and it doesn't work with any Sequence). I could just add `self.data: List` but I thought it is more readable to explicitly make it an empty list.
> 
> To explain the issue, let me explain what is happening to `self.image_data` first. A `sample` which is one entry of `self.image_data` is being passed to `_evaluate_path_location` to not only create `self.data` but also to get enriched by new information:
> 
> https://github.com/Project-MONAI/MONAI/blob/f7f3461b84e55963ef49fd7926825fc1d5c1cad5/monai/data/wsi_datasets.py#L394-L398
> 
> 
> The second part that is adding new information to `sample` depends on `self.image_data` being a mutable so that changing `sample` updates `self.image_data`. This issue was that if the input `data` is immutable, `self.image_data` becomes immutable and updating `sample` not longer updates `self.image_data`.

could you please make a bug report to properly document this issue?

Reply:
Hi @drbeh ,

OK, I see your point. But I think `list(data)` is a shallow copy, if every item in `data` is immutable, it still can't update?
Or maybe I misunderstand your problem? It's better to give a simple example as @wyli suggested.

And I would suggest to remove the `self.data = []`, it's a little bit confusing, `self.data: List` seems clear, we have some similar examples in the codebase:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/array.py#L319
What do you think?

Thanks in advance.

Reply:
Hi @Nic-Ma, you are right! It should address the immutability in each elements. It is fixing the issue that I have encountered when using another datasets as input but I have not defined the problem correctly. I will update the issue and the PR. Thanks for your constructive feedback.

Reply:
Hi @Nic-Ma and @wyli,
I have updated the issue #4724 with more description and an example. From that example it should be clearer why adding a `list` here can resolve the issue. I also updated the PR. Please let me know if you have any question. 
Thanks